### PR TITLE
Support of SSL protocol

### PR DESCRIPTION
**Pull request by oleg-jukovec**
The patch adds support for using SSL to encrypt the client-server
communications [1]. It uses a wrapper around the OpenSSL library for
full compatibility with Tarantool Enterprise (GOST cryptographic
algorithms [2] are not supported by Golang's crypto/tls). The feature
can be disabled using a build tag [3] 'go_tarantool_ssl_disable'.

1. https://www.tarantool.io/ru/enterprise_doc/security/#enterprise-iproto-encryption
2. https://github.com/gost-engine/engine
3. https://pkg.go.dev/go/build#hdr-Build_Constraints

Related issues:

#155 

We have forked go-openssl library (I hope that temporarily). It will be great if you review my commints there too:
https://github.com/tarantool/go-openssl/tree/v0.0.7-24-g0fadeb4-tarantool

I am not familiar with GitHub's CI/CD. I might have missed something. Please, check CI/CD more carefully if you can.

P.S. I will update a Tarantool Enterprise version in tests from beta after the release (on Tuesday?).

### CODE REVIEW

File: ssl_test.go
Comment (ligurio) [inline ssl_test.go:1]: It would be nice to add an example with connection via TLS.

File: .github/workflows/testing.yml
Comment (ligurio) [inline .github/workflows/testing.yml:109]: With your changes, we have two jobs for code coverage. I suppose Coveralls will show a coverage for a job (Community or EE) that post data last. It is not good. In ideal world, we need to measure code coverage for both editions and somehow show both separately on Coveralls.

File: ssl_certs/localhost.crt
Comment (ligurio) [inline ssl_certs/localhost.crt:1]: Usually TLS certificates have expiration dates and some day they will expire. It would be good to describe somewhere steps to generate new certificates.

File: .github/workflows/testing.yml
Comment (ligurio) [inline .github/workflows/testing.yml:9]: > The feature can be disabled using a build tag [3] 'go_tarantool_ssl_disable'.

I suppose it must be disabled for testing of Community Edition. Otherwise, tests will fail.

**Review by ligurio — CHANGES_REQUESTED** [review]

File: .github/workflows/testing.yml
Comment (DifferentialOrange) [inline .github/workflows/testing.yml:90]: I think it's better to use `source tarantool-enterprise/env.sh` where needed

File: ssl_test.go
Comment (ligurio) [inline ssl_test.go:218]: I would also add tests with Russian ciphers too (Stribog, Kuznyechik and probably other Ghost ciphers).

**Review by ligurio — CHANGES_REQUESTED** [review]

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:9]: It's okey. Additional tests with Tarantool EE are run only if a TEST_TNT_EE environment variable exists. Without this, local library integration tests are run, without any Tarantool instance.

File: ssl_test.go
Comment (oleg-jukovec) [reply ssl_test.go:218]: I don't think so. We will test OpenSsl configuration in this case, not our code.

That is a reason why I did not test all ciphers that should work with Tarantool Enterprise. I tested only 1 cipher, which is likely to be almost everywhere.

But maybe I should add a comment about it? Something like:
```
The list of available ciphers depends on the OpenSSL configuration and is not implemented in our code.
```

File: .github/workflows/testing.yml
Comment (ligurio) [reply .github/workflows/testing.yml:9]: Okay.

Probably we should describe the option for our users somewhere in documentation.

File: ssl_test.go
Comment (ligurio) [reply ssl_test.go:218]: > I don't think so. We will test OpenSsl configuration in this case, not our code.

It is not so. We need a simple test to check that it is possible to use Ghost ciphers in Go connector for Tarantool. There is a high probability that our Russian customers will use Ghost ciphers with Tarantool and  I would check it.

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:109]: Thank you. Right now we have only one go-tarantool build: for CE and EE. EE test = CE test + some tests. I see three ways:

1. The result with EE, we assume that it is the most complete.
2. Merge all results.
3. Separate result for each build (Tarantool CE 1.10, Tarantool CE 2.8, Tarantool CE 2.9, Tarantool EE 2.10 etc)

I prefer 2 (if possible) or 1. What do you think?

File: ssl_test.go
Comment (oleg-jukovec) [reply ssl_test.go:218]: This is the reason why we use OpenSSL wrapper. You can add GOST ciphers to OpenSSL, but not at the application level. If we don't try to enable/disable OpenSSL's engines in our application (and we don't), then the engines configured at the system level will work.

I can configure Openssl on CI/CD like that:
https://github.com/tarantool/roadmap-internal/issues/120#issuecomment-1098897905
and test the work of GOST ciphers. But:

1. We will check just that OpenSSL can be configured according to documentation. Nothing about our library.
2. It will be hard to mantain.

I suggest not to claim that we support (or unsupport) any cipher. Because in fact we don't. All we guarantee is working with ciphers that available for OpenSSL with TLSv1.2 inside user's OS.

Comment (oleg-jukovec) [reply ssl_test.go:218]: if we will provide static build with OpenSSL inside, then it would need to be tested. But we don't.

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml]: Thank you, replaced.

File: ssl_test.go
Comment (ligurio) [reply ssl_test.go:218]: Okay, I mark it as resolved.

File: ssl_certs/localhost.crt
Comment (Totktonada) [reply ssl_certs/localhost.crt:1]: I agree that steps to re-generate certificates should be described. I can share my old intructions how to generate testing certificates:

```shell
$ cat deploy/ssl_ca/README
# Gen CA key (password protected) & cert
openssl genrsa -aes256 -out ca.key 4096
chmod 400 ca.key
openssl req -new -x509 -sha256 -key ca.key -out ca.crt -days 365 

# Gen client key & CSR (can be performed on a client side)
openssl genrsa -out client-foo.key 4096
chmod 400 client-foo.key
openssl req -new -key client-foo.key -sha256 -out client-foo.csr

# Gen cert for a client
openssl x509 -req -sha256 -in client-foo.csr -CA ca.crt -CAkey ca.key -set_serial 1 -out client-foo.crt -days 365 
```

```shell
$ cat deploy/ssl_server/README 
# Gen server key (no password to using in nginx) & cert
openssl req -x509 -sha256 -nodes -newkey rsa:4096 -keyout server.key -out server.crt -days 365 
```

```
$ cat deploy/ssl_client/README 
# Gen client key, CSR, and cert for ${SERVICE_NAME}
openssl genrsa -out ${SERVICE_NAME}.key 4096
chmod 400 ${SERVICE_NAME}.key
openssl req -new -key ${SERVICE_NAME}.key -sha256 -out ${SERVICE_NAME}.csr
openssl x509 -req -sha256 -in ${SERVICE_NAME}.csr -CA ../ssl_ca/ca.crt -CAkey ../ssl_ca/ca.key -set_serial 1 -out ${SERVICE_NAME}.crt -days 365 
```

However I don't recommend to generate one-year certificates. It was my mistake. It is quite annoying to find the testing unexpectedly broken because of that.

File: ssl_test.go
Comment (oleg-jukovec) [reply ssl_test.go:1]: Thank you, added. See example_test.go/ExampleSslOpts().

File: ssl_certs/localhost.crt
Comment (oleg-jukovec) [reply ssl_certs/localhost.crt:1]: Thank you, added. See ssl_certs/generate.sh

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:9]: Added, see changes in CONTRIBUTING.md

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:113]: I chose the first one because merging is not supported by goveralls.
(but I can search for alternatives)

File: example_test.go
Comment (DifferentialOrange) [inline example_test.go:28]: Shouldn't there be any restrictions? What happens when we pass SSL options while connecting to a Tarantool without SSL support?

File: example_test.go
Comment (DifferentialOrange) [inline example_test.go:41]: It would be nice to print `err` here too.

File: ssl_certs/generate.sh
Comment (DifferentialOrange) [inline ssl_certs/generate.sh:1]: > testdata... (is a folder to) hold ancillary data needed by the tests

[doc](https://pkg.go.dev/cmd/go/internal/test)

I think it is better to use this structure.

File: ssl.go
Comment (DifferentialOrange) [inline ssl.go:41]: Is there any difference between `return nil, err` on line 33 and `return` here and below? They seem identical to me consider the code,

File: export_test.go
Comment (DifferentialOrange) [inline export_test.go:14]: Isn't go `_test` packages have access to private (lowercase) methods?

**Review by DifferentialOrange — COMMENTED** [review]
To be honest, I'm not familiar with SSL, but since a similar task in assigned to me in tarantool-python, I decided to review that one. Thank you for providing useful links in the description, it helped me to better understand what's going on.

I have left several comments. There are some parts of code I haven't reviewed in-depth yet, but I think I'll give it a shot later.

The main question is compatibility. I've got a bit confused while reviewing the code. What will happen if we connect with SSL-builded connector to older Tarantool? With SSL opts? What do we need ssl_disable dummy functions for: is it only a question of language imports or there is something more?

File: export_test.go
Comment (oleg-jukovec) [reply export_test.go:14]: No, it's kind of valid hack. So you can to export internals, see https://go.dev/src/net/http/export_test.go . But it's better to avoid it.

File: example_test.go
Comment (oleg-jukovec) [reply example_test.go:28]: I think no, it's just an example for godoc. It does not run with tests ( becase no "// Output:" comment).

The example will panic with an error if you run it mannually and SSL is not supported by Tarantool. But it's ok. It will panic with an invalid port too.

File: ssl.go
Comment (oleg-jukovec) [reply ssl.go:41]: Thank you, fixed.

File: ssl_certs/generate.sh
Comment (oleg-jukovec) [reply ssl_certs/generate.sh]: Yes, thank you!

File: example_test.go
Comment (oleg-jukovec) [reply example_test.go]: Thank you, added.

File: CONTRIBUTING.md
Comment (oleg-jukovec) [reply CONTRIBUTING.md]: Yes, updated.

File: config.lua
Comment (oleg-jukovec) [reply config.lua]: Thank you, it works!

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml]: Renamed to TEST_TNT_SSL.

Comment (oleg-jukovec) [reply .github/workflows/testing.yml]: Thank you, updated.

**Comment by oleg-jukovec** [comment]
> The main question is compatibility. I've got a bit confused while reviewing the code. What will happen if we connect with SSL-builded connector to older Tarantool? With SSL opts?

I don't see any problem.

If Transport = 'ssl' is set It will be the same as with another incorrect configuration (host, port etc). Just do it in a correct way (don't set Transport = 'ssl', set up right ip, port etc) and all be fine.

> What do we need ssl_disable dummy functions for: is it only a question of language imports or there is something more?

An user will be able to make build without OpenSSL dependency (maybe he doesn't need it), as it was before. Linking with the OpenSSL library can complicate build.

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml]: 1.10, 2.8 added

> and it looks like it would be better to separate that commit from SSL support introduction

Ok, i'll fetch changes inside .github/ into a separate commi before merge.

File: ssl.go
Comment (oleg-jukovec) [reply ssl.go:1]: Yes, we agreed that SSL should be enabled by default. Using the tag, we disable SSL support. It is the reason why such name is used.

A tag can be set to almost any go command:
```bash
go build -tags tag_name
go run -tags tag_name
go test -tags tag_name
```
https://pkg.go.dev/go/build#hdr-Build_Constraints

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:103]: No problems, updated.

File: export_test.go
Comment (oleg-jukovec) [reply export_test.go]: I think it would be better to delete the obvious comment (we don't want to share the internal implementation, it's a hack etc). It's just confusing and has no real value. I deleted it.

File: test_helpers/main.go
Comment (DifferentialOrange) [inline test_helpers/main.go:46]:
```suggestion
	// PingSsl changes Opts.Ssl for a connection that checks startup
```

Comment (oleg-jukovec) [reply test_helpers/main.go]: Thank you!

File: ssl_test.go
Comment (oleg-jukovec) [reply ssl_test.go]: Half of the tests are integration tests with OpenSSL, without a Tarantool instance. Half of the tests are integration tests with Tarantool.

I renamed *Local* to *Ssl* and *Ee* to *Tnt* in the tests file. Maybe this will make the difference more obvious.

File: testdata/generate.sh
Comment (ligurio) [inline testdata/generate.sh:2]: please add `set -e`

File: testdata/generate.sh
Comment (ligurio) [inline testdata/generate.sh:1]: Probably it is worth to integrate script with Makefile or even move command to it. However it's up to you.

File: ssl_test.go
Comment (ligurio) [reply ssl_test.go:1]: Thanks!

File: ssl_certs/localhost.crt
Comment (ligurio) [reply ssl_certs/localhost.crt]: (it was renamed to `testdata/generate.sh`)

File: ssl_test.go
Comment (ligurio) [inline ssl_test.go:1]: We have a single connection to a single instance and so-called multiconnection. Shouldn't we add a small test for muticonnection with SSL?

**Review by ligurio — CHANGES_REQUESTED** [review]

File: testdata/generate.sh
Comment (oleg-jukovec) [reply testdata/generate.sh:1]: I've added a ```testdata``` target into Makefile
```patch
+.PHONY: testdata
+testdata:
+       (cd ./testdata; ./generate.sh)
+
```

File: ssl_test.go
Comment (oleg-jukovec) [reply ssl_test.go:1]: I don't think so:

Multiconnection is about handling and managing multiple connections. Not about a single connection and it options. It would be too much to test the connection options there (for example, we don't test in multi_test.go other connection options, just multiconnection logic).

File: .github/workflows/testing.yml
Comment (Totktonada) [reply .github/workflows/testing.yml:9]: (I'll resolve the discussion: it looks fixed by the note in CONTRIBUTING.md.)

File: .github/workflows/testing.yml
Comment (Totktonada) [reply .github/workflows/testing.yml:119]: I'm ok with the first approach.

Comment (Totktonada) [reply .github/workflows/testing.yml:119]: (Resolving the discussion so.)

File: testdata/generate.sh
Comment (Totktonada) [reply testdata/generate.sh:1]: Seems resolved, so I'll resolve the discussion.

File: testdata/generate.sh
Comment (Totktonada) [reply testdata/generate.sh]: You wrote `test -e`.

My standard preambles are the following (depending on a situation):

```shell
#!/bin/sh

set -eu # Strict shell (w/o -o pipefail)
```

```shell
#!/bin/sh

set -xeu # Strict shell (w/o -o pipefail)
```

```shell
#!/bin/bash

set -xeuo pipefail # Strict shell
```

Motivated by http://redsymbol.net/articles/unofficial-bash-strict-mode/

I don't like playing with `IFS`, however. It feels dangerous.

File: testdata/generate.sh
Comment (Totktonada) [inline testdata/generate.sh:25]: If possible, I would generate testing certificates without an expiration date.

File: .github/workflows/testing.yml
Comment (Totktonada) [inline .github/workflows/testing.yml:6]: Nit: I would replace `/ru/` in the link in the commit message (and PR description) with `/en/` to make it readable for everyone.

File: ssl.go
Comment (Totktonada) [reply ssl.go:1]: I would discuss this point once again. One of possible reasons to use Go in various tools is portability of the resulting executable. Linking with particular version of openssl may decrease portability of the executable. So maybe it should be explicit choice of a user.

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:6]: Sure, fixed.

File: testdata/generate.sh
Comment (oleg-jukovec) [reply testdata/generate.sh]: I can't find a way to do it. I've Increased the number of days to 8192. That should be enough:
```bash
$ openssl x509 -enddate -noout -in ca.pem 
notAfter=Oct 29 06:17:40 2044 GMT
$ openssl x509 -enddate -noout -in localhost.crt 
notAfter=Oct 29 06:17:40 2044 GMT
```

File: testdata/generate.sh
Comment (oleg-jukovec) [reply testdata/generate.sh]: Thank you. I chose the last one. Updated.

File: ssl.go
Comment (oleg-jukovec) [reply ssl.go:1]: It seems to me that it is better to provide this feature by default. This is a new functionality that we plan to support and provide. This is not an alternative to something that already exists. I don't see any reason from the library why it should be disabled.

Not every user of the library will use the feature, but the problem with linking and portability is not too difficult to notice. And then add a tag to disable it if needed.

My conclusion: from the point of view of the library, this should be enabled by default. From the user's point of view, he can easily disable it.

Comment (Totktonada) [reply ssl.go:1]: 'Easy to change' sometimes is not so easy. Say, software vendor A ships an application build to company B, but the operation team is actually in company C (B's integrator). Shipping a new application version means some agreement between all three on an upgrade process. It may involve a lot of people: developers, managers and operation team from different sides. Well, there may be also a pure technical complexity: the application has a component, which has a component, which actually has the problem and the software vendor needs to fix the nested component, release and deploy it, update the outer component, release and deploy it, update the application, release it and give the built to B&C. All those processes are performed by different teams (in A) with its own SLA between them.

So time between the point when one found that something is broken and actual fix may be quite large. Not taking even taking into account amount of effort of different people to do so.

(This picture is not very different from what we actually have between us and our customers.)

The risk of problem is not always compensated by simplicity of the change.

I'm not insist on staying on the most safe side here, but, please, take this picture into account. Since there is no strong arguments for enabling or disabling (only personal feeling of 'risky' and 'easy to change'), I'll leave it up to the author of the patch. Closing the discussion.

File: ssl_test.go
Comment (Totktonada) [reply ssl_test.go:1]: The connection pool (as well as the deprecated multi API) just passes the options to individual connections, so it looks orthogonal enough to don't insist on a special test. Closing the discussion.

File: testdata/generate.sh
Comment (Totktonada) [reply testdata/generate.sh]: Okay, thanks!

File: test_helpers/main.go
Comment (Totktonada) [reply test_helpers/main.go]: Seems resolved. Closing the discussion.

File: ssl_test.go
Comment (Totktonada) [reply ssl_test.go]: Seems resolved. Closing the discussion.

**Review by Totktonada — APPROVED** [review]
LGTM. I don't see problems in the code and can't add more value here.

Well, there is (a low?) risk of extra dynamic dependency on system's openssl (discussed in Georgy's comment). I'm okay to let Oleg to take the risk.

I would like to know more regarding ability to link openssl statically into resulting executable and leave some instructions for a user. Maybe even a sanity check like 'if static-openssl build flag is enabled, the connector will fail on building at attempt to link it to openssl dynamically'. However I'm okay to work on it later if there will be a demand.

Let's acquire a second approve (I think Georgy can finish its review) and proceed.

File: .github/workflows/testing.yml
Comment (DifferentialOrange) [inline .github/workflows/testing.yml:115]: Shouldn't we have `env: TEST_TNT_SSL: ${{matrix.ssl}}` here too?

File: connection.go
Comment (DifferentialOrange) [inline connection.go:420]: If we're missing libssl, libcrypto and never use `transport == connTransportSsl`, will application with go-tarantool work is it was built with default flags? I'm not well up in linking.

File: ssl_test.go
Comment (DifferentialOrange) [inline ssl_test.go:148]: Code below `Fatalf` is not executed. Is it what we really want?

File: ssl_test.go
Comment (DifferentialOrange) [inline ssl_test.go:165]: Not `Fatalf`?

File: ssl_test.go
Comment (DifferentialOrange) [inline ssl_test.go:178]: Not `Fatalf`? [2]

File: test_helpers/main.go
Comment (DifferentialOrange) [inline test_helpers/main.go:240]: There is no reason to choose a server inside the cycle on each iteration

**Review by DifferentialOrange — COMMENTED** [review]
Should be LGTM after resolving remaining things.

To be honest, I'm also not a fan of adding an SSL dependencies by default, but if discussions result is like this, well, I will not oppose it.

File: test_helpers/main.go
Comment (oleg-jukovec) [reply test_helpers/main.go]: Thank you. It's a good idea to replace `Ping` because it's a implementation detail. So I've renamed `PingServer`, `PingTransport`, `PingSsl` to `ClientServer`, `ClientTransport`, `ClientSsl`.

It useful to set these values directly in tests (to set the wrong ones, for example, we use SSL for a server, but not for a client).

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:65]: Agree, added.

File: .github/workflows/testing.yml
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:116]: Oh, I missed it. Thank you, fixed!

File: ssl_test.go
Comment (oleg-jukovec) [reply ssl_test.go:148]: Yes, you are right. Thank you, updated.

File: test_helpers/main.go
Comment (oleg-jukovec) [reply test_helpers/main.go]: Agree, the code has been extracted from the loop.

File: ssl_test.go
Comment (oleg-jukovec) [reply ssl_test.go:165]: I don't return anything from `assert*` functions and do a clean up inside. `Errorf` is enough. Anyway, I've replaced all `Fatalf` with `Errorf` in `assert*` functions to make it less confusing.

File: connection.go
Comment (Totktonada) [reply connection.go:420]: The resulting executable will have the dynamic dependency on libssl and libcrypto (shown in `ldd` or `otool -L`). The dynamic loader will refuse to load a dynamic executable if any library is missing.

It is safe while given executable is used within a distro of particular version (because distros guarantees ABI compatibility -- where we had openssl 1.1, it will not be updated to openssl 3; at least not from base repositories). So if you build an application for each distro or at least have separate builds for openssl 1.1 and openssl 3 distros, everything should be fine.

If you want to build an executable that would work on, say, Fedora 35 and Fedora 36 both (or Ubuntu Impish and Ubuntu Jammy both), you're in the trouble.

So the question is actually about ways to deliver the resulting application. I can't judge about some 'usual way' here.

**Comment by Totktonada** [comment]
> <...> maybe we should leave this patch until the next tag, after we tag current stable and compatible things?

Okay. It seems, it is time to [make a release](https://github.com/tarantool/go-tarantool/wiki/How-to-release). @oleg-jukovec, can you, please, release the connector?

**Comment by oleg-jukovec** [comment]
> @oleg-jukovec, can you, please, release the connector?

Sure, [it's done](https://github.com/tarantool/go-tarantool/releases/tag/v1.6.0).

**Pull request merged**
